src/data/list/basic.lean

```diff
@@ -3164,6 +3164,74 @@ by convert filter_eq_nil.2 (λ _ _, id)
       take_while_append_drop l]
     else by rw [take_while, drop_while, if_neg pa, if_neg pa, nil_append]
 
+lemma drop_while_nth_le_zero_not (l : list α) (hl : 0 < (l.drop_while p).length) :
+  ¬ p ((l.drop_while p).nth_le 0 hl) :=
+begin
+  induction l with hd tl IH,
+  { rw length_pos_iff_exists_mem at hl,
+    contrapose! hl,
+    simp [drop_while] },
+  { simp only [drop_while],
+    split_ifs with hp,
+    { exact IH _ },
+    { simpa using hp } }
+end
+
+variables {p} {l : list α}
+
+@[simp] lemma drop_while_eq_nil_iff : drop_while p l = [] ↔ ∀ x ∈ l, p x :=
```

Comment:
I don't have time to try to golf right now, but this looks a lot more convoluted than I'd expect a proof of this theorem to be. Are you sure it can't be cleaned up?